Moving UserStorage

ahus1 opened the pull request:
…reManager and MapUserProvider

<!---
Please read https://github.com/keycloak/keycloak/blob/master/CONTRIBUTING.md and follow these guidelines when contributing to Keycloak
-->

ahus1 committed:
WIP: Moving UserStorage, some commented out code in UserCredentialStoreManager and MapUserProvider

hmlnarik commented on model/legacy-private/src/main/java/org/keycloak/storage/datastore/LegacyMigrationManager.java:
```diff
@@ -0,0 +1,181 @@
+    }
+
+    @Override
+    public void migrate(RealmModel realm, RealmRepresentation rep, boolean skipUserDependent) {
```
seems not to match org.keycloak.migration.MigrationModelManager.migrateImport(KeycloakSession, RealmModel, RealmRepresentation, boolean)?

ahus1 replied:
should be resolved now.

hmlnarik commented on model/map-ldap/src/main/java/org/keycloak/models/map/storage/ldap/config/LdapMapConfig.java:
```diff
@@ -271,12 +270,12 @@ public boolean isStartTls() {
     }
 
-    public UserStorageProvider.EditMode getEditMode() {
+    public EditMode getEditMode() {
```
Why has been `EditMode` introduced rather than leveraging `UserStorageProvider.EditMode`?

hmlnarik commented on model/map/src/main/java/org/keycloak/models/map/user/MapUserProvider.java:
```diff
@@ -454,9 +453,12 @@ public void preRemove(ClientScopeModel clientScope) {
         String componentId = component.getId();
         LOG.tracef("preRemove[ComponentModel](%s, %s)%s", realm, componentId, getShortStackTrace());
+        /*
+        TODO: why?
```
Because the map user provider used to be used with legacy stores for other areas. I believe it can be removed now.

ahus1 replied:
removed.

hmlnarik commented on server-spi-private/src/main/java/org/keycloak/storage/MigrationManager.java:
```diff
@@ -0,0 +1,14 @@
+/**
+ * @author Alexander Schwartz
+ */
+public interface MigrationManager {
```
Should vanish in the future.

ahus1 replied:
Added JavaDoc, will keep for now.

hmlnarik commented on server-spi/src/main/java/org/keycloak/storage/user/UserBulkUpdateProvider.java:
```diff
@@ -21,7 +21,7 @@
 /**
  * This is an optional capability interface that is intended to be implemented by any
- * {@link org.keycloak.storage.UserStorageProvider UserStorageProvider} that supports bulk operations.
+ * {@link org.keycloak.models.UserProvider} that supports bulk operations.
```
`UserProvider` implements this interface (`UserBulkUpdateProvider`).
Same for all other interfaces in this package

ahus1 replied:
just use code instead of a link

ahus1 replied:
Updated JavaDoc with just the code

hmlnarik commented on model/legacy-private/src/main/java/org/keycloak/migration/migrators/MigrateTo1_4_0.java:
```diff
@@ -55,7 +55,8 @@ protected void migrateRealm(KeycloakSession session, RealmModel realm) {
         ImpersonationConstants.setupImpersonationService(session, realm);
 
-        migrateLDAPMappers(session, realm);
+        // TODO: changed the order of execution here ... does that matter?
```
It seems ok in this case

ahus1 replied:
Alex will revert.

ahus1 replied:
cleaned up.

ahus1 committed:
fixup! WIP: Moving UserStorage, some commented out code in UserCredentialStoreManager and MapUserProvider

ahus1 commented:
@hmlnarik - Status update: cleanup of the bits and pieces done, pushed in a separate commit to the PR's branch. Import and Update now handled by the ExportImportManager, with larger portions of RepresentationToModel moved to the legacy module. Will start moving over code for the UserCredentialStoreManager tomorrow.

ahus1 committed:
SingleUserCredentialManager moving in

ahus1 committed:
fixup! SingleUserCredentialManager moving in

Now handling login via LDAP + Kerberos

ahus1 committed:
Moving classes

ahus1 committed:
Introducing authenticate() method on the realm, and moving code to both old and new store

ahus1 commented:
@hmlnarik - pushed two changes: 
1. moved classes around DefaultSingleUserCredentialManager
2. Realm has now a method authenticate() - with implementations for JPA and Map storage. I was able to log in locally with LDAP + Kerberos as an interactive test.

Looking forward to discuss more on Monday.

hmlnarik commented on model/map/src/main/java/org/keycloak/models/map/realm/MapRealmAdapter.java:
```diff
@@ -1761,6 +1767,26 @@ public void decreaseRemainingCount(ClientInitialAccessModel model) {
+                .map(credentialAuthentication -> credentialAuthentication.authenticate(this, input))
+                .filter(Objects::nonNull)
+                .findFirst().orElse(null);
+    }
```
I understand this mimics the legacy behaviour, which is to loop over all `CredentialProvider`s available in the system, and as such is good for the intermediate state of this PR.

In the final version of the PR, this needs to be handled differently: The obtained `CredentialProvider`s need to be relevant for this realm only. It still may be that it would be all of those registered in the system, but not necessarily so.

Since the result of the credential authentication is a `UserModel` (in `CredentialValidationOutput.authenticatedUser`), this seems to be leveraging the `getUserByCredential` method in `UserProvider` and calling (`session.users().getUserByCredential(realm, input);`).

Inside `UserProvider.getUserByCredential(RealmModel, CredentialInput)`:
1. Checks:
    - The obtained [storage](https://github.com/hmlnarik/keycloak/blob/0c9ab32cf4dc116c9134effd40d86f3351e6da32/model/map/src/main/java/org/keycloak/models/map/user/MapUserProvider.java#L84) (`store`) would be checked to implement a newly introduced interface `AuthenticatingStorage` with a single added method `boolean supportsCredentialType(String credentialType)` and one overridden method `AuthenticatingTransaction<V> createTransaction(KeycloakSession)`, and
    - The obtained [transaction](https://github.com/hmlnarik/keycloak/blob/0c9ab32cf4dc116c9134effd40d86f3351e6da32/model/map/src/main/java/org/keycloak/models/map/user/MapUserProvider.java#L86) would be checked to implement a newly introduced interface `AuthenticatingTransaction<V>` with a new method `V authenticate(CredentialInput)`. It behaves similarly to `read(id)` operation, only determining the user from `CredentialInput` rather than `id`.
    - Note: This would replace the added `authenticate(RealmModel realm, CredentialInput input)` method from [`MapKeycloakTransaction`](https://github.com/hmlnarik/keycloak/pull/23/files#diff-f5ad0c497c1429f1259a831ebc77ef332d613b85fce55be7ce85171284080a93)

2. If the storage supports credential type and the transaction implements `AuthenticatingTransaction` interface, then the `authenticate` method is called.

The main difference is that it is not the `MapRealmAdapter` but the storage transaction that decides which of the providers to try. This could be tree storage (virtual) transaction which delegates the `authenticate` method down the tree accordingly.

hmlnarik commented on model/legacy-private/pom.xml:
```diff
@@ -32,6 +32,11 @@
             <groupId>org.keycloak</groupId>
             <artifactId>keycloak-model-legacy</artifactId>
         </dependency>
+        <dependency>
+            <groupId>org.jboss.resteasy</groupId>
+            <artifactId>resteasy-jaxrs</artifactId>
+            <scope>provided</scope>
+        </dependency>
```
This would be expected in legacy-services. Is it necessary here?

ahus1 replied:
I received a complaint in the IDE about the LegacyExportImportManager that is calling ValidationUtil.validateClient (which is in server-spi-private) might throw a BadRequestException exception, which is in that package. Maybe we can discuss tomorrow how to handle this. If something needs to move, it would be LegacyExportImportManager + LegacyMigrationManager.

```
/home/aschwart/redhat/workspace/keycloak-pr-1/model/legacy-private/src/main/java/org/keycloak/storage/datastore/LegacyExportImportManager.java:437:42
java: cannot access javax.ws.rs.BadRequestException
  class file for javax.ws.rs.BadRequestException not found
```

hmlnarik replied:
Thank you, let's keep this as is, even though this is not ideal. Perhaps a new exception would be better than this.

Could you please add a comment to this place in `pom.xml` referring this comment thread?

hmlnarik commented on model/legacy-private/src/main/java/org/keycloak/credential/DefaultSingleUserCredentialManagerStrategy.java:
```diff
@@ -0,0 +1,105 @@
+        if (StorageId.isLocalStorage(user.getId())) {
+            return (UserCredentialStore) session.userLocalStorage();
+        } else {
+            return (UserCredentialStore) session.userFederatedStorage();
```
This has to target `datastore`, not `session`: `session.userLocalStorage()` will be equivalent to `session.users()` [like in other cases](https://github.com/hmlnarik/keycloak/blob/3ad0669d0f28cbc14934400dfa073aca9adab5a3/services/src/main/java/org/keycloak/services/DefaultKeycloakSession.java#L173-L196), and so will `session.userFederatedStorage()` after this refactor.

ahus1 replied:
At the moment during legacy mode `session.userLocalStorage()` will return the local storage, while `session.users()` will return the cache or the federated store. Some callers seems to rely on that semantic, I'd like to have a look at this together as I'm hesistant to change the DefaultKeycloakSession's implementation.

hmlnarik commented on model/legacy-private/src/main/java/org/keycloak/storage/datastore/LegacyDatastoreProvider.java:
```diff
@@ -184,4 +193,15 @@ public UserProvider users() {
+    }
+
+    @Override
+    public MigrationManager getMigrationManager() {
```
This will vanish, please add deprecation note

hmlnarik commented on model/map/src/main/java/org/keycloak/models/map/credential/SingleUserCredentialManagerEntity.java:
```diff
@@ -0,0 +1,33 @@
+import java.util.List;
+
+/**
+ * Interface for credential management in entities in the map storage.
```
Please add javadoc to the methods.

hmlnarik commented on model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/entities/CachedUser.java:
```diff
@@ -119,4 +124,10 @@ public Set<String> getGroups(Supplier<UserModel> userModel) {
     }
+
+    public List<CredentialModel> getStoredCredentials(Supplier<UserModel> userModel) {
+        // TODO: should the list and its items be cloned first to avoid side effects?
```
No, credential manipulation is done by special methods, and cloning `CredentialModel` is nontrivial.

ahus1 replied:
yes, clone it to be safe

ahus1 replied:
will be cloned in next commit

hmlnarik commented on model/jpa/src/main/java/org/keycloak/models/jpa/UserAdapter.java:
```diff
@@ -515,6 +517,11 @@ public void setServiceAccountClientLink(String clientInternalId) {
 
+    @Override
+    public SingleUserCredentialManager getUserCredentialManager() {
+        return session.getProvider(SingleUserCredentialManagerProvider.class).create(realm, this);
```
Would this work if `SingleUserCredentialManagerProvider` was a map user credential manager?
I believe this should be a concrete implementation rather than provider lookup, i.e.
```suggestion
        return new DefaultSingleUserCredentialManager(realm, this, new DefaultSingleUserCredentialManagerStrategy(session, realm, user));
```

hmlnarik commented on model/legacy-private/src/main/java/org/keycloak/credential/DefaultSingleUserCredentialManagerProvider.java:
```diff
@@ -0,0 +1,57 @@
+/**
+ * Creates an instance of the SingleUserCredentialManager.
+ *
+ * There is only one implementation at the moment. It uses the SPI as {@link DefaultSingleUserCredentialManager}
```
Would there ever be more than one implementation per user datastore?

hmlnarik commented on model/map/src/main/java/org/keycloak/models/map/user/MapUserAdapter.java:
```diff
@@ -265,7 +265,8 @@ public void leaveGroup(GroupModel group) {
     @Override
     public boolean isMemberOf(GroupModel group) {
-        return RoleUtils.isMember(getGroupsStream(), group);
+        Set<String> groups = entity.getGroupsMembership();
+        return groups != null && groups.contains(group.getId());
```
`isMemberOf` accounts for hierarchy. This change modifies the semantics, and is unrelated to legacy store. Please fix.

hmlnarik commented on model/map/src/main/java/org/keycloak/models/map/user/MapUserEntity.java:
```diff
@@ -246,4 +247,14 @@ public Boolean removeFederatedIdentity(String identityProviderId) {
     void setNotBefore(Long notBefore);
+
+    /**
+     * Validate the provided credentials for this user.
```
This Javadoc seems obsolete

ahus1 replied:
updated, please resolve.

hmlnarik commented on model/map/src/main/java/org/keycloak/models/map/user/MapUserProvider.java:
```diff
@@ -102,6 +105,11 @@ public boolean checkEmailUniqueness(RealmModel realm, String email) {
+
+            @Override
+            public SingleUserCredentialManager getUserCredentialManager() {
+                return session.getProvider(SingleUserCredentialManagerProvider.class).create(realm, this, new MapSingleUserCredentialManagerStrategy(entity));
```
What would happen if `SingleUserCredentialManagerProvider` provider was a unrelated one? (same as above)

hmlnarik commented on server-spi-private/src/main/java/org/keycloak/credential/SingleUserCredentialManagerProviderFactory.java:
```diff
@@ -0,0 +1,27 @@
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.provider.ProviderFactory;
+
+public interface SingleUserCredentialManagerProviderFactory extends ProviderFactory<SingleUserCredentialManagerProvider> {
```
I tend to think that this interface should vanish

ahus1 replied:
Changed to a concrete implementation in the next commit and it is called LegacySingleUserCredentialManager. The accompanying LegacySingleUserCredentialManagerStrategy will eventually be folded into that class but not yet to make it simpler to compare with the MapSingleUserCredentialManager

hmlnarik commented on server-spi-private/src/main/java/org/keycloak/migration/MigrationModelManager.java:
```diff
@@ -17,169 +17,24 @@
  * @author <a href="mailto:bill@burkecentral.com">Bill Burke</a>
  * @version $Revision: 1 $
  */
 public class MigrationModelManager {
```
Please add deprecation note that this class will be moved

ahus1 replied:
added in next commit

hmlnarik commented on server-spi/src/main/java/org/keycloak/credential/SingleUserCredentialManagerProvider.java:
```diff
@@ -0,0 +1,35 @@
+ * Interface to create an instance of a {@link SingleUserCredentialManager}, either with a default strategy or
+ * a custom strategy.
+ */
+public interface SingleUserCredentialManagerProvider extends Provider {
```
I am not convinced this should be a provider. Let's discuss it on Monday

ahus1 replied:
This has now been removed.

hmlnarik reviewed (commented):
Thanks @ahus1, this looks good. I've left a few comments / questions, please find them inline

ahus1 committed:
fixup! Moving classes

ahus1 committed:
Avoiding a provider for DefaultSingleUserCredentialManager via factory, introducing variant for map storage. UserStorageManager now handles authentication for old Kerberos+LDAP style, new getUserByCredential in MapUserProvider would eventually do the same.

ahus1 committed:
adding deprecation

ahus1 committed:
removing obsolete classes around SingleUserCredentialManagerSpi

ahus1 commented:
@hmlnarik - I committed some (most?) changes we discussed yesterday. I was able to manually login with the legacy store, the map store and with legacy LDAP+Kerberos.

ahus1 committed:
renaming and moving of classes, fixing dependencies

ahus1 commented:
@hmlnarik - changes of today's meeting pushed including moving of LegacySingleUserCredentialManager. On order to move that class, also AbstractStorageManager needed to be moved.

hmlnarik commented on model/map/src/main/java/org/keycloak/models/map/storage/MapStorageWithAuth.java:
```diff
@@ -0,0 +1,19 @@
+ */
+public interface MapStorageWithAuth<V extends AbstractEntity & UpdatableEntity, M> extends MapStorage<V, M> {
+
+    boolean supportsCredentialAuthenticationFor(String type);
```
Can this be renamed (for consistency sake) to `supportsCredentialType`?

ahus1 replied:
renamed

hmlnarik reviewed (commented):
Thanks for the update!

The wildfly failures are caused by missing entries in `module.xml` of keycloak-model-jpa (and likely -infinispan) module, see https://github.com/hmlnarik/keycloak/commit/fc52adbde40558dfdf7dbb1fb4b21cb98c491c6e

hmlnarik commented on model/map/src/main/java/org/keycloak/models/map/user/MapCredentialValidationOutput.java:
```diff
@@ -0,0 +1,63 @@
+ *
+ * @author <a href="mailto:mposolda@redhat.com">Marek Posolda</a>
+ */
+public class MapCredentialValidationOutput {
+
+    private final MapUserEntity authenticatedUser; // authenticated user.
```
Just realized that this needn't be `MapUserEntity` in case of partial stores which do not necessarily return the entity but only an intermediate object which is turned into a `MapUserEntity` by mappers.

Could you please replace this with
```suggestion
public class MapCredentialValidationOutput<V> {

    private final V authenticatedUser; // authenticated user.
```

This can be safely cast to `MapUserEntity` in the [`getUserByCredentials` method](https://github.com/hmlnarik/keycloak/pull/23/files#diff-90c66f213572f75c2455b1b176eb7b16c3740209567dc7e5002253f882852f67R788) - partial store would by that time have it converted to `MapUserEntity` by the mappers.

ahus1 replied:
generified, see c0bf1d58f29eef149f9a4b6f69c364d5640c2c3e

ahus1 committed:
updating why this dependency is needed

ahus1 committed:
renaming classes, making them generic

ahus1 committed:
avoid using methods on UserCredentialStoreManager, step 1

ahus1 committed:
avoid using methods on UserCredentialStoreManager, step 2

ahus1 committed:
avoid using methods on UserCredentialStoreManager, step 3

ahus1 committed:
avoid using methods on UserCredentialStoreManager, step 4

ahus1 committed:
avoid using methods on UserCredentialStoreManager, step 5

(inlining)

ahus1 committed:
avoid using methods on UserCredentialStoreManager, step 6

ahus1 committed:
avoid using methods on UserCredentialStoreManager, step 7

ahus1 committed:
avoid using methods on UserCredentialStoreManager, step 8

ahus1 committed:
avoid using methods on UserCredentialStoreManager, step 9

ahus1 committed:
avoid using methods on UserCredentialStoreManager, step 10

ahus1 committed:
avoid using methods on UserCredentialStoreManager, step 11

ahus1 committed:
avoid using methods on UserCredentialStoreManager, step 12

ahus1 committed:
avoid using methods on UserCredentialStoreManager, step 13

ahus1 committed:
avoid using methods on UserCredentialStoreManager, step 14

ahus1 commented:
@hmlnarik - just pushed 14 steps to avoid using UserCredentialStoreManager in non-legacy-mode. As expected, this touched a lot of classes, and I wonder if this should really be part of the PR as it makes it difficult to review.

ahus1 committed:
fixup! Users

Adding new SPI to the Spi file to be discovered

hmlnarik commented on model/map/src/main/java/org/keycloak/models/map/user/MapCredentialValidationOutput.java:
```diff
@@ -27,23 +27,23 @@
-    public static MapCredentialValidationOutput failed() {
-        return new MapCredentialValidationOutput(null, CredentialValidationOutput.Status.FAILED, new HashMap<String, String>());
+    public static MapCredentialValidationOutput<?> failed() {
+        return new MapCredentialValidationOutput<Void>(null, CredentialValidationOutput.Status.FAILED, new HashMap<>());
```
`Collections.emptyMap()` might be better here than `new HashMap<>()`

hmlnarik commented on model/map/src/main/java/org/keycloak/models/map/credential/MapSingleUserCredentialManagerEntity.java:
```diff
@@ -0,0 +1,33 @@
+ * @author Alexander Schwartz
+ */
+public interface MapSingleUserCredentialManagerEntity {
+    void validateCredentials(List<CredentialInput> inputs);
+
+     boolean updateCredential(CredentialInput input);
```
Please add javadoc to the methods.

ahus1 committed:
Disabling tests for LDAP/Kerberos that won't work when map storage is enabled

ahus1 committed:
fixup! Users

ahus1 committed:
Added LegacySessionSupport SPI

While some methods around onCache() are still called from the legacy code, all other methods log a warning with a stacktrace.

ahus1 committed:
fixup! Disabling tests for LDAP/Kerberos that won't work when map storage is enabled

ahus1 committed:
Deprecated old APIs

ahus1 committed:
Renaming of API without "get" prefix to make it consistent with other APIs like for example with KeycloakSession

ahus1 committed:
fixup! Disabling tests for LDAP/Kerberos that won't work when map storage is enabled

ahus1 committed:
fixup! Added LegacySessionSupport SPI

ahus1 commented:
@hmlnarik - pushed some changes yesterday: Tests are now all green. Added LegacySessionSupport SPI.

Will continue on Monday with polishing, especially JavaDoc and your recent comments.

ahus1 committed:
updated JavaDoc for new classes and methods

ahus1 committed:
renamed class SingleUserCredentialManager to SingleEntityCredentialManager, and method userCredentialManager to credentialManager

ahus1 commented:
@hmlnarik - completed rename to credentialManager and SingleEntityCredentialManager.
Pushed the changes to the draft branch, waiting for your merge.
(local tests executed: login with jpa map store and legacy h2 store, clean install runs successfully)
